File: Jim/src/sk/fiit/jim/agent/highskill/Localize.java
package sk.fiit.jim.agent.highskill;

/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.models.DynamicObject;
import sk.fiit.jim.agent.models.EnvironmentModel;
import sk.fiit.jim.agent.models.WorldModel;
import sk.fiit.jim.agent.moves.LowSkill;
import sk.fiit.jim.agent.moves.LowSkills;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Angles;

/**
 *
 * Reimplemented from Ruby to Java by roman moravcik and stefan linner
 */
public class Localize extends HighSkill {

    private Angles right = new Angles(180.0, 360.0);
    private Angles left = new Angles(0.0, 180.0);
    private Angles front1 = new Angles(0.0, 90.0);
    private Angles front2 = new Angles(270.0, 360.0);
    private Angles back = new Angles(90.0, 270.0);
    
    private boolean leftLook = false;
    private boolean rightLook = false;

    
    public Localize(){};
    
    public void checkProgress() {
    }

    public LowSkill pickLowSkill() {
        DynamicObject ball = WorldModel.getInstance().getBall();
        double ballAngle = Angles.normalize(ball.getRelativePosition().getPhi());
        AgentModel agentModel = AgentModel.getInstance();
        
        if (EnvironmentModel.beamablePlayMode()) {
            return null;
        } 
        else if (agentModel.isOnGround()) {
            return null;
        }
        else if (!leftLook) {
            leftLook = true;
            return LowSkills.get("head_left_120");
        }
        else if (!rightLook) {
            rightLook = true;
            return LowSkills.get("head_right_120");
        }
        else if (WorldModel.getInstance().isSeeBall() ) {
        	return null;
        }
        else if (left.include(ballAngle) && (front1.include(ballAngle) || front2.include(ballAngle))) {
            // Vlavo vpredu
            return LowSkills.get("turn_left_cont_20");
        } 
        else if (right.include(ballAngle) && (front1.include(ballAngle) || front2.include(ballAngle))) {
            // Vpravo vpredu
            return LowSkills.get("turn_right_cont_20");
        }
        else if (left.include(ballAngle) && back.include(ballAngle)) {
            // Vlavo vzadu
            return LowSkills.get("turn_left_90");
        }
        else if (right.include(ballAngle) && back.include(ballAngle)) {
            // Vpravo vzadu
            return LowSkills.get("turn_right_90");
        }

        return LowSkills.get("turn_right_90");
    }
}

File: Old_xml_and_scripts/Jim/high_skills/old-predUcesanim/localize.rb
require 'high_skills/ruby_high_skill.rb'
include_class Java::sk.fiit.robocup.library.geometry.Angles
include_class Java::java.lang.Math

class Localize < RubyHighSkill

  @@right = 180..360
  @@left = 0..180
  @@front1 = 0..90
  @@front2 = 270..360
  @@back = 90..270
  
  @@go1 = 0..30
  @@go2 = 330..360
  
  @leftLook = false;
  @rightLook = false;
  
  def initialize validity_proc
      super()
      @validity_proc = validity_proc
      @agentInfo = Java::sk.fiit.jim.agent.AgentInfo.getInstance
      @agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance
      
      @leftLook = false;
      @rightLook = false;
  end
  
  def pickLowSkill
    @ball = WorldModel.getInstance.getBall.getRelativePosition
    case
          when (not @validity_proc.call) : 
            puts "validity"
            return nil
          when EnvironmentModel.beamablePlayMode :
            puts "beam" 
            return nil
          when @agentModel.falled? :
            puts "fall" 
            return nil
      
          when !@leftLook :
            puts "left look"
            @leftLook = true
            return get_skill("head_left_120")
          when !@rightLook :
            puts "right look"
            @rightLook = true
            return get_skill("head_right_120")
              
                 
         when (ballInRange? @@left and (ballInRange? @@front1 or ballInRange? @@front2)) :
           puts "vlavo vpredu" 
           return get_skill("turn_left_cont_20")
         when (ballInRange? @@right and (ballInRange? @@front1 or ballInRange? @@front2)) :
           puts "vpravo vpredu"  
           return get_skill("turn_right_cont_20")
         
         when (ballInRange? @@left and ballInRange? @@back) :
           puts "vlavo vzadu"  
           return get_skill("turnleft90_faster_4")
         when (ballInRange? @@right and ballInRange? @@back) :
           puts "vpravo vzadu"  
           return get_skill("turnright90_faster_4")    
       end 
      return get_skill("turnright90_faster_4")    
  end
 
  
  def ballInRange? range
    range.include? (Math.toDegrees(@ball.phi))
  end  
  
  def checkProgress
    #do nothing
  end
    
end	
#in order not to defer the startup cost
Localize.new nil